### ISSUE TITLE
Geth can't estimate gas for the keyless deployment

### ISSUE THREAD

**Issue by ChihChengLiang** [thehubbleproject/hubble-contracts#533]
### What's wrong

If we run the deploy script on Geth (1.9.25 and 1.10.0), this line fails.

https://github.com/thehubbleproject/hubble-contracts/blob/ff3e64cacfdaf2c22cdc3a46ad91b3eff8fd3626/ts/deployment/keylessDeployment.ts#L54-L56

With this error message

```
(node:28608) UnhandledPromiseRejectionWarning: Error: insufficient funds for intrinsic transaction cost (error={"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"id\":46,\"error\":{\"code\":-32000,\"message\":\"insufficient funds for transfer\"}}\n","error":{"code":-32000},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"gas\":\"0x5da38\",\"gasPrice\":\"0x174876e800\",\"value\":\"0x0\",\"from\":\"0xe2856e9ab0c0258d9a3c70c444c99b43e0a8b607\",\"data\":\"0x608060405234801561001057600080fd5b5061049c806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806332c02a141461003b578063c090587f14610083575b600080fd5b6100676004803603604081101561005157600080fd5b506001600160a01b0381351690602001356100a0565b604080516001600160a01b039092168252519081900360200190f35b6100676004803603602081101561009957600080fd5b503561018f565b60006100b36100ae836101a0565b61020c565b156100fe576040805162461bcd60e51b815260206004820152601660248201527511195c1b1bde595c8e881cd85b1d081a5cc81d5cd95960521b604482015290519081900360640190fd5b6060610108610248565b9050828151602083016000f59150813b61012157600080fd5b816001600160a01b031663bd5b2202856040518263ffffffff1660e01b815260040180826001600160a01b03168152602001915050600060405180830381600087803b15801561017057600080fd5b505af1158015610184573d6000803e3d6000fd5b505050505092915050565b600061019a826101a0565b92915050565b600060ff30836101ae610248565b80519060200120604051602001808560ff1660f81b8152600101846001600160a01b031660601b81526014018381526020018281526020019450505050506040516020818303038152906040528051906020012060001c9050919050565b6000813f7fc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47081811480159061024057508115155b949350505050565b60606040518060200161025a90610272565b601f1982820381018352601f90910116604052905090565b610210806102808339019056fe608060405234801561001057600080fd5b506101f0806100206000396000f3fe6080604052600436106100295760003560e01c80630fd5b6ae14610033578063bd5b220214610064575b610031610097565b005b34801561003f57600080fd5b506100486100b1565b604080516001600160a01b039092168252519081900360200190f35b34801561007057600080fd5b506100316004803603602081101561008757600080fd5b50356001600160a01b03166100c0565b61009f6100af565b6100af6100aa61019a565b6101bf565b565b60006100bb61019a565b905090565b60006100ca61019a565b6001600160a01b031614610125576040805162461bcd60e51b815260206004820152601a60248201527f50726f78793a20616c726561647920696e697469616c697a6564000000000000604482015290519081900360640190fd5b6001600160a01b038116610176576040805162461bcd60e51b815260206004820152601360248201527250726f78793a207a65726f206164647265737360681b604482015290519081900360640190fd5b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c355565b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c35490565b3660008037600080366000845af43d6000803e8080156101de573d6000f35b3d6000fdfea164736f6c634300060c000aa164736f6c634300060c000a\"}],\"id\":46,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"http://localhost:8545"}, method="estimateGas", transaction={"from":"0xE2856e9ab0C0258d9a3C70c444C99B43e0a8B607","gasLimit":{"type":"BigNumber","hex":"0x05da38"},"gasPrice":{"type":"BigNumber","hex":"0x174876e800"},"value":{"type":"BigNumber","hex":"0x00"},"data":"0x608060405234801561001057600080fd5b5061049c806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806332c02a141461003b578063c090587f14610083575b600080fd5b6100676004803603604081101561005157600080fd5b506001600160a01b0381351690602001356100a0565b604080516001600160a01b039092168252519081900360200190f35b6100676004803603602081101561009957600080fd5b503561018f565b60006100b36100ae836101a0565b61020c565b156100fe576040805162461bcd60e51b815260206004820152601660248201527511195c1b1bde595c8e881cd85b1d081a5cc81d5cd95960521b604482015290519081900360640190fd5b6060610108610248565b9050828151602083016000f59150813b61012157600080fd5b816001600160a01b031663bd5b2202856040518263ffffffff1660e01b815260040180826001600160a01b03168152602001915050600060405180830381600087803b15801561017057600080fd5b505af1158015610184573d6000803e3d6000fd5b505050505092915050565b600061019a826101a0565b92915050565b600060ff30836101ae610248565b80519060200120604051602001808560ff1660f81b8152600101846001600160a01b031660601b81526014018381526020018281526020019450505050506040516020818303038152906040528051906020012060001c9050919050565b6000813f7fc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47081811480159061024057508115155b949350505050565b60606040518060200161025a90610272565b601f1982820381018352601f90910116604052905090565b610210806102808339019056fe608060405234801561001057600080fd5b506101f0806100206000396000f3fe6080604052600436106100295760003560e01c80630fd5b6ae14610033578063bd5b220214610064575b610031610097565b005b34801561003f57600080fd5b506100486100b1565b604080516001600160a01b039092168252519081900360200190f35b34801561007057600080fd5b506100316004803603602081101561008757600080fd5b50356001600160a01b03166100c0565b61009f6100af565b6100af6100aa61019a565b6101bf565b565b60006100bb61019a565b905090565b60006100ca61019a565b6001600160a01b031614610125576040805162461bcd60e51b815260206004820152601a60248201527f50726f78793a20616c726561647920696e697469616c697a6564000000000000604482015290519081900360640190fd5b6001600160a01b038116610176576040805162461bcd60e51b815260206004820152601360248201527250726f78793a207a65726f206164647265737360681b604482015290519081900360640190fd5b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c355565b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c35490565b3660008037600080366000845af43d6000803e8080156101de573d6000f35b3d6000fdfea164736f6c634300060c000aa164736f6c634300060c000a"}, code=INSUFFICIENT_FUNDS, version=providers/5.0.24)
    at Logger.makeError (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/logger/src.ts/index.ts:205:28)
    at Logger.throwError (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/logger/src.ts/index.ts:217:20)
    at checkError (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:48:16)
    at JsonRpcProvider.<anonymous> (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:463:20)
    at step (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:48:23)
    at Object.throw (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:29:53)
    at rejected (/home/ubuntu/hubble-contracts/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:21:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:28608) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:28608) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```

parsed tx

```
parsed tx {
  nonce: 0,
  gasPrice: BigNumber { _hex: '0x174876e800', _isBigNumber: true },
  gasLimit: BigNumber { _hex: '0x05da38', _isBigNumber: true },
  to: null,
  value: BigNumber { _hex: '0x00', _isBigNumber: true },
  data: '0x608060405234801561001057600080fd5b5061049c806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806332c02a141461003b57
8063c090587f14610083575b600080fd5b6100676004803603604081101561005157600080fd5b506001600160a01b0381351690602001356100a0565b604080516001600160a01b03909216825251908
1900360200190f35b6100676004803603602081101561009957600080fd5b503561018f565b60006100b36100ae836101a0565b61020c565b156100fe576040805162461bcd60e51b8152602060048201
52601660248201527511195c1b1bde595c8e881cd85b1d081a5cc81d5cd95960521b604482015290519081900360640190fd5b6060610108610248565b9050828151602083016000f59150813b6101215
7600080fd5b816001600160a01b031663bd5b2202856040518263ffffffff1660e01b815260040180826001600160a01b03168152602001915050600060405180830381600087803b1580156101705760
0080fd5b505af1158015610184573d6000803e3d6000fd5b505050505092915050565b600061019a826101a0565b92915050565b600060ff30836101ae610248565b80519060200120604051602001808
560ff1660f81b8152600101846001600160a01b031660601b81526014018381526020018281526020019450505050506040516020818303038152906040528051906020012060001c9050919050565b60
00813f7fc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47081811480159061024057508115155b949350505050565b60606040518060200161025a90610272565b601f198
2820381018352601f90910116604052905090565b610210806102808339019056fe608060405234801561001057600080fd5b506101f0806100206000396000f3fe608060405260043610610029576000
3560e01c80630fd5b6ae14610033578063bd5b220214610064575b610031610097565b005b34801561003f57600080fd5b506100486100b1565b604080516001600160a01b03909216825251908190036
0200190f35b34801561007057600080fd5b506100316004803603602081101561008757600080fd5b50356001600160a01b03166100c0565b61009f6100af565b6100af6100aa61019a565b6101bf565b
565b60006100bb61019a565b905090565b60006100ca61019a565b6001600160a01b031614610125576040805162461bcd60e51b815260206004820152601a60248201527f50726f78793a20616c72656
1647920696e697469616c697a6564000000000000604482015290519081900360640190fd5b6001600160a01b038116610176576040805162461bcd60e51b815260206004820152601360248201527250
726f78793a207a65726f206164647265737360681b604482015290519081900360640190fd5b7f7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c355565b7f7050c9e0f4c
a769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c35490565b3660008037600080366000845af43d6000803e8080156101de573d6000f35b3d6000fdfea164736f6c634300060c000aa164
736f6c634300060c000a',
  chainId: 0,
  v: 27,
  r: '0x0001abb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1e',
  s: '0x0002abb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1eabb1e',
  from: '0xE2856e9ab0C0258d9a3C70c444C99B43e0a8B607',
  hash: '0xce4b180a1bac535f2d742a36460cbe1b1040c500491cdaf203b5ac2f05d6d176'
}
```

### How can we fix it?

No idea yet

**Comment by ChihChengLiang** [comment]
Workaround: if we fund this address "0xE2856e9ab0C0258d9a3C70c444C99B43e0a8B607" the estimate_gas can work.

But in the production environment, not sure if this workaround is desirable.

**Comment by ChihChengLiang** [comment]
Addressed in #536. 

When doing keyless deployment, we need to fund the keyless account 0xE2856 before deploying the contract.

But before funding the keyless account and deploying the contract, we had a check that makes sure the gas limit is high enough to deploy the contract, which requires the estimate_gas for the raw tx, which Geth requires the sender of raw tx to be funded first.

In #536, we disable the gas limit check.

Why do we have this check in the first place? We have the following constraints to satisfy, so the gas limit needs to be hardcoded and checked before the actual deployment.

- We hardcode the BN pairing cost estimator's address in the BLS contract. The address must remain constant.
- The BN pairing cost estimator's address is generated by Create2, which depends on 3 things:
  - On-chain create2 deployer's address
  - A salt
  - The hash of the proxy contract's bytecode
- The on-chain create2 deployer is deployed by the keyless deployer. Its contract address is derived from the account that deployed it.
- The keyless deployer's account is derived from the signature and the hash of the RLP encoded deployment tx.
- The signature is constant, but to get a constant hash of tx, we need tx fields to be constant. 
- That means gasPrice and gasLimit need to be constant.
